### DISCUSSION TITLE
Welcome to LinuxCNC-RIO Discussions!

### DISCUSSION THREAD

**Discussion by multigcs** [multigcs/LinuxCNC-RIO#11] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by EhsanRasouli1371** [comment]
I want to build a 6 axis closed loop controller using Colorlight5A-75E  board and a hp mini pc and six servo motors. Can anyone guide me how to do this?

**Comment by multigcs** [comment]
the Colorlight5A-75E Ethernet-Port is not supported by RIO
and servos not really tested :(

better take a look at https://github.com/Peter-van-Tol/LiteX-CNC

**Reply by EhsanRasouli1371** [reply]
I used to use Litex.
But I have a problem with it and unfortunately Peter has been inactive for a while.
Can you solve the problems I have raised there?

**Comment by EhsanRasouli1371** [comment]
Can I implement your work on this development board which has spartan 6 processor?
I want to have 6 closed loop axes
![برد-mojo-plus](https://github.com/multigcs/LinuxCNC-RIO/assets/147668561/a454689b-3d7d-4592-aa28-4a3ed5b1723d)

**Reply by multigcs** [reply]
https://forum.linuxcnc.org/27-driver-boards/51052-spartan-6-lx9-dev-board-as-mesa-7c81

**Comment by EhsanRasouli1371** [comment]
Do you think it is possible to put the firmware related to the Mesa boards on the Colorlight5A-75E board?

**Reply by ghost** [reply]
I've been looking in at this discussion for a bit and I thought I give some insight.
For starters this is what I came up with.
https://forum.linuxcnc.org/show-your-stuff/51142-spartan-6-fpga-card-9d60

Some points of note:
The DIY route is not the cheapest way to do it, and can be frustrating at times, especially when it comes to the FPGA bit stream side.
This project maybe of use it uses the Etherent interface and works with the hm2_eth driver. Any thing to do hal wise can be found on the Linuxcnc forums, it wont be specific to this board but as it uses the Linuxcnc hostmot drivers the information should be easy to find.

Out of all the Spartan 6 Dev boards I found this one had a good number of available pins and the form factor was good to work with.
I have played with some of the Spartan  Colorlight boards but they require some mods to the board to get flexibility and then they 3.3 to 5v conversion needs looking at. I did manage to get one going with an SPI interface and one smart serial port, the smart serial port required extra hardware.
Cost is also another factor, whilst the dev & colorlight boards seem cheap and they are, designing & build the hardware to work with 5v & 24v signal levels can be not the easiest task but also expensive. If you want to get a machine up & running with the least amount of troubles the Mesa infrastructure is the way to go. If you have the bits & pieces available it's not so bad.  
Pete & et all this is no slight or a write off of your projects, I'm quite impressed with all the work & time you guys put in.
There is quite a bit of information on the Linuxcnc forums about "rolling your own Mesa card", there's a lot to digest and just blindly using someone else's firmware doesn't give you an insight into how it all works and trouble shooting.
Getting the clock setup is a little tricky but not so bad, what really was the biggest headache was working out which pins on which bank to use for the SPI interface was the difficult part. If the clock signal does not coincide with the Mesa implementation it's time to start reading through the Xilinx data sheets.
Whilst working on my projects I spent close to what would have been the cost of a 7i92 delivered to Australia. 
Costs to consider are:
The initial purchase of the dev board, the Mojo boards are crazy expensive. Even the one I used wasn't the cheapest but was what I considered the most versatile.
What JTag programmer are you going to use, do you have to buy one ?
The cost of getting PCB boards to suit (and any mistakes in design), unless you are going to use veroboard or the like.
The cost and availability of buying the required components, you usually end up buying in bulk. 
Also do you want to go all SMD of a mixture of through hole & SMD. After soldering the SMD chips I was glad the rest were though hole.
Time spent getting the FPGA bitstreams right and testing, this takes a while.
It's also worth studying some of the Mesa schematics that are available for download on the forum, this can help give some ideas on the electrical side of things. Full 7c81 & 7i92 are available as well as some partial schematics.
Most of the time there is "no one to hold your hand" you need to do alot of the grunt work yourself. It helps to have good experience with electronics and programming, no getting a LED to flash on an Arduino with code that you copied is not the kind of experience I'm talking about. It's the experience you get from starting projects from scratch.
I may have gone a bit over the top with the project I linked to, but I wanted to make it as versatile as possible, hence the reason for the daughter boards, using the same dev board and daughter boards I can use either a SPI interface or EPP interface with appropriate bitsreams. I also had to mod the Linuxcnc drivers to get full use of the available pins when not using 7c81 firmware. I did think about an Ethernet daughter board, but that was a bit over the top. Another thing to I wanted my boards to be able to be used with either a standard Parallel Port break out board or any of the MESA db25 daughter cards.
Another thing to consider if you do have an issue posting on the forum saying that you have built your own Mesa type board and are having issues, whether HAL or something else may not get you a lot on answers.
Sheesh I have even posted some fixes to the Linuxcnc 2.9.2 amd64 iso and asked for help in testing, the only "help" I got was a user cloning from the Linuxcnc Live build repo, at the time my changes hadn't been merged, and and saying it doesn't work. I did post a link to my repo with the fixes, which have since been merged. Even after posting a link to the ISO no-one has made any comments. 
Sorry for the long winded post & whinge but there is a lot to think about. I feel Peter has been most patient with your questions.

**Reply by EhsanRasouli1371** [reply]
Thank you very much for taking the time to share your experience with us.
I also agree with you, but the problem is that I don't have access to Mesa boards regardless of their price to buy.
The only way for me is to either make them or find something similar.
Another thing is that I unfortunately do not have the experience of programming FPGA boards and currently I have to use existing projects.
Also, I am very grateful to people like you and Peter

**Comment by ghost** [comment]
What is the issue with getting a Mesa board ?

**Reply by EhsanRasouli1371** [reply]
I live in Iran and due to sanctions, I cannot do financial transactions with European countries, Canada and America.

**Comment by ghost** [comment]
Oh ok I get it.

There is a Mesa supplier in Portugal https://eusurplus.com/index.php?route=information/contact

And a list here:
https://forum.linuxcnc.org/27-driver-boards/34979-mesa-suppliers#114946

There is also one in New Zealand, I'm not too sure about the financial restrictions. 

Look mate I live in Australia if there is someway to get a board to you I could help. I've got my email on my github account, you can send me an email and we can sort something out if you want.
Cheers
Rob
